### PR TITLE
Implement func `array_pop_front`

### LINKED ISSUE #6996: Implement `array_pop_front` function
## Summary

| Characteristic        | Description                                                                                                                                                               |
| --------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Function name:        | `array_pop_front`                                                                                                                                                         |
| Aliases:              | `list_pop_front`                                                                                                                                                          |
| Original function?:   | No                                                                                                                                                                        |
| Function Description: | <b>ClickHouse:</b> Removes the first item from the array. </br> <b>DuckDB:</b> Returns the list without the first element.                                                |
| Sources:              | [Concept](https://github.com/apache/arrow-datafusion/discussions/6855) [DuckDB](https://duckdb.org/docs/sql/functions/nested) [ClickHouse](https://clickhouse.com/docs/en/sql-reference/functions/array-functions#arraypopfront) |

## Examples:

```
D select array_pop_front([1, 2, 3, 4]);
┌──────────────────────────────────────────────┐
│ array_pop_front(main.list_value(1, 2, 3, 4)) │
│                   int32[]                    │
├──────────────────────────────────────────────┤
│ [2, 3, 4]                                    │
└──────────────────────────────────────────────┘
D select array_pop_front([[1, 2], [3, 4], [4, 2]]);
┌──────────────────────────────────────────────────────────────────────────────┐
│ array_pop_front(main.list_value(main.list_value(1, 2), main.list_value(3, …  │
│                                  int32[][]                                   │
├──────────────────────────────────────────────────────────────────────────────┤
│ [[3, 4], [4, 2]]                                                             │
└──────────────────────────────────────────────────────────────────────────────┘
```

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Veeupup**
## Which issue does this PR close?

Closes #6996 .

### CODE REVIEW

**Comment by Veeupup** [comment]
@alamb @jayzhan211  hi! this PR is ready for review now! : )

**Comment by jayzhan211** [comment]
@Veeupup Would you like to try rewriting array_pop_front / back with MutableArrayData? I think it will be much easier to understand than the current slice! and list_slice! macros.

File: datafusion/physical-expr/src/array_expressions.rs [hunk 1]
```diff
@@ -563,13 +563,42 @@ pub fn array_slice(args: &[ArrayRef]) -> Result<ArrayRef> {
     define_array_slice(list_array, key, extra_key, false)
 }
 
+fn general_array_pop(
+    list_array: &GenericListArray<i32>,
+    from_back: bool,
+) -> Result<(Vec<i64>, Vec<i64>)> {
+    if from_back {
```
Comment (jayzhan211) [inline datafusion/physical-expr/src/array_expressions.rs:570]: nit: Since there is no common for back and front, I dont think we need to write them together, but that's fine.

**Review by jayzhan211 — APPROVED** [review]

**Comment by Veeupup** [comment]
> @Veeupup Would you like to try rewriting array_pop_front / back with MutableArrayData? I think it will be much easier to understand than the current slice! and list_slice! macros.

yeah, agree with you.. to understand what slice! macro does cost me some time..

but this macro has been used for some other functions and for now implementation with `slice!` maybe the simplest way to do with it?

**Comment by jayzhan211** [comment]
> > @Veeupup Would you like to try rewriting array_pop_front / back with MutableArrayData? I think it will be much easier to understand than the current slice! and list_slice! macros.
> 
> yeah, agree with you.. to understand what slice! macro does cost me some time..
> 
> but this macro has been used for some other functions and for now implementation with `slice!` maybe the simplest way to do with it?

If you also spend much time on slice!, it is a good reason for us to improve on this.

**Review by comphead — COMMENTED** [review]
Thanks @Veeupup 
Please more tests with nulls and empty arrays, please try to think up any possible case to prevent bug to flow in.

And since new built in function gets introduced we need to get this documented in `scalar_functions.md` and `expressions.md`

File: datafusion/physical-expr/src/array_expressions.rs [hunk 2]
```diff
@@ -563,13 +563,42 @@ pub fn array_slice(args: &[ArrayRef]) -> Result<ArrayRef> {
...
+        let key = vec![0; list_array.len()];
+        let extra_key: Vec<_> = list_array
+            .iter()
+            .map(|x| x.map_or(0, |arr| arr.len() as i64 - 1))
+            .collect();
+        Ok((key, extra_key))
+    } else {
+        let key: Vec<_> = list_array.iter().map(|x| x.map_or(0, |_arr| 2)).collect();
```
Comment (comphead) [inline datafusion/physical-expr/src/array_expressions.rs:578]: what is 2 here? and we dont use `arr` we probably can just put `||`

Comment (Veeupup) [reply datafusion/physical-expr/src/array_expressions.rs:578]: 2 is the first element position (position = index + 1) we want in the new array.

We only handle two cases of the first element index: if the old array has any elements, starts from 2 (index + 1), or starts from initial.

**Comment by Veeupup** [comment]
> Thanks @Veeupup Please more tests with nulls and empty arrays, please try to think up any possible case to prevent bug to flow in.
> 
> And since new built in function gets introduced we need to get this documented in `scalar_functions.md` and `expressions.md`

very thanks for your advice! docs have been added!

and the test for empty and NULL has already added, and you can check [empty array test](https://github.com/apache/arrow-datafusion/pull/8142/files#diff-317c67cc9ce87268e4ccec1cb75316eed82f99ae2ffc226874e5897913ffa4c8R838) and [NULL value test](https://github.com/apache/arrow-datafusion/pull/8142/files#diff-317c67cc9ce87268e4ccec1cb75316eed82f99ae2ffc226874e5897913ffa4c8R850).

File: datafusion/physical-expr/src/array_expressions.rs (hunk 2 above)
Comment (comphead) [reply datafusion/physical-expr/src/array_expressions.rs:579]: thanks, please leave the comment in the code what the 2 is, so for other devs it will be easier to read the code.

File: datafusion/physical-expr/src/array_expressions.rs [hunk 3]
```diff
@@ -564,13 +564,30 @@ pub fn array_slice(args: &[ArrayRef]) -> Result<ArrayRef> {
...
```
Comment (comphead) [inline datafusion/physical-expr/src/array_expressions.rs:579]:
```suggestion
        let key: Vec<_> = list_array.iter().map(|x| x.map_or(0, |_| 2)).collect();
```

File: datafusion/physical-expr/src/array_expressions.rs (hunk 2 above)
Comment (Veeupup) [reply datafusion/physical-expr/src/array_expressions.rs]: Fixed, thanks again! : )

**Review by alamb — APPROVED** [review]
Thanks @Veeupup
